coriolis/api/v1/replicas.py

```diff
@@ -109,6 +110,135 @@ def delete(self, req, id):
+        for mapping in backend_mappings:
+            if mapping['source'] not in new_backend_mapping_sources:
+                non_duplicates_backend_mapping.append(mapping)
+            else:
+                LOG.info("Storage Backend Mapping %s will be overwritten." %
+                         mapping)
+
+        non_duplicates_disk_mappings = []
+        for mapping in disk_mappings:
+            if mapping['disk_id'] not in new_disk_mappings_disk_ids:
+                non_duplicates_disk_mappings.append(mapping)
+            else:
+                LOG.info("Storage Disk Mapping %s will be overwritten" %
+                         mapping)
+
+        storage_mappings = {
+            'backend_mappings': non_duplicates_backend_mapping.extend(
+                new_backend_mappings),
+            'disk_mappings': non_duplicates_disk_mappings.extend(
+                new_disk_mappings)}
+
+        default_storage_backend = (
+            new_storage_mappings.get('default', None) or
+            original_storage_mappings.get('default', None))
+        if default_storage_backend:
+            storage_mappings['default'] = default_storage_backend
+
+        return storage_mappings
+
+    def _get_merged_replica_values(self, replica, updated_values):
+        final_values = {}
+        # NOTE: this just replaces options at the top-level and does not do
+        # merging of container types (ex: lists, dicts)
+        for option in [
+            "source_environment", "destination_environment", "network_map"]:
+            original_values = replica.get(option, {})
+            final_values[option] = original_values.update(updated_values.get(
+                option, {}))
+
+        original_storage_mappings = replica.get('storage_mappings', {})
+        new_storage_mappings = updated_values.get('storage_mappings', {})
+        final_values['storage_mappings'] = self._update_storage_mappings(
+            original_storage_mappings, new_storage_mappings)
+
+        if 'notes' in updated_values:
+            final_values['notes'] = updated_values.get('notes', '')
+        else:
+            final_values['notes'] = replica.get('notes', '')
+
+        # until the provider plugin interface is updated
+        # to have separate 'network_map' and 'storage_mappings' fields,
+        # we add them as part of the destination environment:
+        final_storage_mappings = final_values['storage_mappings']
+        final_network_map = final_values['network_map']
+        if final_storage_mappings:
+            final_values[
+                'destination_environment'][
+                'storage_mappings'] = final_storage_mappings
+        if final_network_map:
+            final_values[
+                'destination_environment'][
+                'network_map'] = final_network_map
+
+        return final_values
+
+    def _validate_update_body(self, id, context, body):
+
+        replica = self._replica_api.get_replica(context, id)
```

Comment:
Please add explicit checks and return an explicit error message is someone tries to update `origin_endpoint_id`, `destination_endpoint_id` or `instances`.